File: ms-pedidos/src/main/java/br/com/florinda/pedidos/domain/Pedido.java
package br.com.florinda.pedidos.domain;

import io.quarkus.hibernate.orm.panache.PanacheEntityBase;
import jakarta.persistence.*;
import jakarta.validation.Valid;
import jakarta.validation.constraints.*;

import java.math.BigDecimal;
import java.time.OffsetDateTime;
import java.util.ArrayList;
import java.util.List;
import java.util.UUID;

@Entity
@Table(name = "pedidos")
public class Pedido extends PanacheEntityBase {

    @Id
    @Column(columnDefinition = "CHAR(36)")
    public UUID id = UUID.randomUUID();

    @NotNull
    @Column(name = "cliente_id", nullable = false, columnDefinition = "CHAR(36)")
    public UUID clienteId;

    @NotNull
    @Column(name = "restaurante_id", nullable = false, columnDefinition = "CHAR(36)")
    public UUID restauranteId;

    @NotNull
    @Enumerated(EnumType.STRING)
    @Column(nullable = false)
    public StatusPedido status = StatusPedido.PENDENTE;

    @NotNull
    @DecimalMin(value = "15.00", message = "Pedido mínimo de R$ 15,00")
    @Column(name = "valor_total", nullable = false, precision = 10, scale = 2)
    public BigDecimal valorTotal;

    @Size(max = 500)
    public String observacao;

    @Size(max = 500)
    @Column(name = "endereco_entrega")
    public String enderecoEntrega;

    @Column(name = "criado_em", nullable = false, updatable = false)
    public OffsetDateTime criadoEm = OffsetDateTime.now();

    @Column(name = "atualizado_em", nullable = false)
    public OffsetDateTime atualizadoEm = OffsetDateTime.now();

    @Valid
    @OneToMany(mappedBy = "pedido", cascade = CascadeType.ALL,
               orphanRemoval = true, fetch = FetchType.EAGER)
    public List<ItemPedido> itens = new ArrayList<>();

    @OneToMany(mappedBy = "pedido", cascade = CascadeType.ALL,
               fetch = FetchType.LAZY)
    public List<HistoricoStatusPedido> historico = new ArrayList<>();

    // -----------------------------------------------------------
    // Comportamentos de domínio (RN02, RN03, RN04, RN05)
    // -----------------------------------------------------------

    public void confirmar() {
        transicionarStatus(StatusPedido.CONFIRMADO, "Pagamento aprovado");
    }

    public void iniciarPreparo() {
        transicionarStatus(StatusPedido.PREPARANDO, "Restaurante iniciou o preparo");
    }

    public void sairParaEntrega() {
        transicionarStatus(StatusPedido.SAIU_PARA_ENTREGA, "Pedido saiu para entrega");
    }

    public void entregar() {
        transicionarStatus(StatusPedido.ENTREGUE, "Pedido entregue ao cliente");
    }

    public void cancelar(String motivo) {
        if (!this.status.podeSerCancelado()) {
            throw new IllegalStateException(
                "Pedido no status " + this.status + " não pode ser cancelado. " +
                "Cancelamento permitido apenas em PENDENTE e CONFIRMADO.");
        }
        transicionarStatus(StatusPedido.CANCELADO, motivo);
    }

    public void falharPagamento() {
        transicionarStatus(StatusPedido.CANCELADO, "Pagamento recusado");
    }

    private void transicionarStatus(StatusPedido destino, String motivo) {
        if (!this.status.podeTransicionarPara(destino)) {
            throw new IllegalStateException(
                "Transição inválida: " + this.status + " → " + destino);
        }
        HistoricoStatusPedido registro = HistoricoStatusPedido.registrar(
                this, this.status, destino, motivo);
        this.historico.add(registro);
        this.status       = destino;
        this.atualizadoEm = OffsetDateTime.now();
    }

    public void calcularTotal() {
        this.valorTotal = itens.stream()
                .map(i -> i.precoUnitario.multiply(BigDecimal.valueOf(i.quantidade)))
                .reduce(BigDecimal.ZERO, BigDecimal::add);
    }

    // -----------------------------------------------------------
    // Queries estáticas
    // -----------------------------------------------------------

    public static List<Pedido> findByCliente(UUID clienteId) {
        return list("clienteId", clienteId);
    }

    public static List<Pedido> findByStatus(StatusPedido status) {
        return list("status", status);
    }

    public static List<Pedido> findByRestaurante(UUID restauranteId) {
        return list("restauranteId", restauranteId);
    }
}

File: mobile/lib/features/pedido/models/pedido.dart
class ItemPedidoRequest {
  final String itemId;
  final String nomeItem;
  final double precoUnitario;
  final int quantidade;

  const ItemPedidoRequest({
    required this.itemId,
    required this.nomeItem,
    required this.precoUnitario,
    required this.quantidade,
  });

  Map<String, dynamic> toJson() => {
        'itemId': itemId,
        'nomeItem': nomeItem,
        'precoUnitario': precoUnitario,
        'quantidade': quantidade,
      };
}

class CriarPedidoRequest {
  final String clienteId;
  final String restauranteId;
  final String enderecoEntrega;
  final String? observacao;
  final List<ItemPedidoRequest> itens;

  const CriarPedidoRequest({
    required this.clienteId,
    required this.restauranteId,
    required this.enderecoEntrega,
    this.observacao,
    required this.itens,
  });

  Map<String, dynamic> toJson() => {
        'clienteId': clienteId,
        'restauranteId': restauranteId,
        'enderecoEntrega': enderecoEntrega,
        if (observacao != null) 'observacao': observacao,
        'itens': itens.map((i) => i.toJson()).toList(),
      };
}

class PedidoResponse {
  final String id;
  final String status;
  final double valorTotal;
  final String? observacao;
  final String enderecoEntrega;
  final DateTime criadoEm;

  const PedidoResponse({
    required this.id,
    required this.status,
    required this.valorTotal,
    this.observacao,
    required this.enderecoEntrega,
    required this.criadoEm,
  });

  factory PedidoResponse.fromJson(Map<String, dynamic> j) => PedidoResponse(
        id: j['id'] as String,
        status: j['status'] as String,
        valorTotal: (j['valorTotal'] as num).toDouble(),
        observacao: j['observacao'] as String?,
        enderecoEntrega: j['enderecoEntrega'] as String,
        criadoEm: DateTime.parse(j['criadoEm'] as String),
      );
}